python/vsi/tools/patch_wheel.py

```diff
@@ -1,77 +1,112 @@
   else:
-    version = file_version
-
-  new_filename = filename.parent / ('-'.join([normalized_name,
-                                              version,
-                                              remainder]) + '.whl')
-
-  with tempfile.NamedTemporaryFile(suffix='.whl',
-                                   prefix='.' + filename.stem + '.',
-                                   dir=filename.parent,
-                                   delete=False) as temp_file:
-    os.chmod(temp_file.name, 0o644) # Python hardcodes this to 600 :(
-    with zipfile.ZipFile(filename, 'r') as zin:
-      with zipfile.ZipFile(temp_file, 'w') as zout:
-        zout.comment = zin.comment # preserve the comment
-        for item in zin.infolist():
-          # Patch METADATA to match filename
-          if item.filename.endswith('.dist-info/METADATA'):
-            meta_lines = (patch[key]
-                          if (key := line.split(b':',1)[0]) in patch.keys()
-                          else line
-                          for line in zin.read(item.filename).split(b'\n'))
-            zout.writestr(item, b'\n'.join(meta_lines))
-          else:
-            zout.writestr(item, zin.read(item.filename))
-  os.rename(temp_file.name, new_filename)
+    version = original_version
+
+  # https://peps.python.org/pep-0491/#escaping-and-unicode
+  normalized_name = re.sub(r"[^\w\d.]+", "_", name).lower()
+  normalized_version = re.sub(r"[^\w\d.+]+", "_", version).lower()
+
+  if name == original_name and version == original_version:
+    raise Exception('No change detected. Package name and version are '
+                    'identical')
+
+  with tempfile.TemporaryDirectory() as temp_extract:
+    # Extract Wheel
+    subprocess.Popen(['wheel', 'unpack', '--dest', temp_extract,
+                      filename]).wait()
+
+    # Patch metadata
+    metadata = (Path(temp_extract) /
+                f'{original_name}-{original_version}' /
+                f'{original_name}-{original_version}.dist-info/METADATA')
+    with open(metadata, 'r+') as fid:
+      meta_lines = (patch[key]
+                    if (key := line.split(':',1)[0]) in patch.keys()
+                    else line
+                    for line in fid.readlines())
+      fid.seek(0)
+      fid.writelines(meta_lines)
+
+
+    # Rename  files
+    os.rename(Path(temp_extract) /
+              f'{original_name}-{original_version}' /
+              f'{original_name}-{original_version}.dist-info',
+              Path(temp_extract) /
+              f'{original_name}-{original_version}' /
+              f'{normalized_name}-{normalized_version}.dist-info')
+    os.rename(Path(temp_extract) /
+              f'{original_name}-{original_version}',
+              Path(temp_extract) /
+              f'{normalized_name}-{normalized_version}')
+
+    # Save new wheel
+    with tempfile.TemporaryDirectory() as temp_wheel:
+      subprocess.Popen(['wheel', 'pack', '--dest', temp_wheel,
```

Comment:
surprised you can't do this from API calls

Reply:
I could try, but I know it's not wise to call CLI modules from python, they aren't designed for that and I don't feel like risking it (`pip` is an example)